File: backend/src/main/java/com/imedba/modules/enrollment/entity/Enrollment.java
package com.imedba.modules.enrollment.entity;

import com.imedba.common.entity.BaseEntity;
import com.imedba.modules.course.entity.Course;
import com.imedba.modules.student.entity.Student;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.Builder.Default;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "enrollments")
@SQLDelete(sql = "UPDATE enrollments SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
public class Enrollment extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "course_id", nullable = false)
    private Course course;

    /** FK a discount_campaigns — la entidad/módulo llega en Fase 2, acá sólo guardamos el id. */
    @Column(name = "discount_campaign_id")
    private UUID discountCampaignId;

    /** sub del JWT de la vendedora que cargó la inscripción. */
    @Column(name = "enrolled_by")
    private UUID enrolledBy;

    @Column(name = "enrollment_date", nullable = false)
    private Instant enrollmentDate;

    @Column(name = "list_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal listPrice;

    @Default
    @Column(name = "discount_percentage", nullable = false, precision = 5, scale = 2)
    private BigDecimal discountPercentage = BigDecimal.ZERO;

    @Column(name = "final_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal finalPrice;

    @Default
    @Column(name = "book_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal bookPrice = BigDecimal.ZERO;

    @Column(name = "total_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal totalPrice;

    @Column(name = "enrollment_fee", precision = 12, scale = 2)
    private BigDecimal enrollmentFee;

    @Default
    @Column(name = "num_installments", nullable = false)
    private Integer numInstallments = 1;

    // El medio de pago vive en la cuota/pago (Payment), no en la inscripción
    // (reunión 2026-06-08). Columna payment_method dropeada en V024.

    @Column(name = "contract_file_path", length = 500)
    private String contractFilePath;

    @Column(name = "contract_sent_at")
    private Instant contractSentAt;

    @Column(name = "contract_signed_at")
    private Instant contractSignedAt;

    @Default
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 30)
    private EnrollmentStatus status = EnrollmentStatus.ACTIVE;

    /**
     * Grupo de pago: define el día de vencimiento de las cuotas (10 o 20) y el umbral del
     * recargo. Reunión 2026-06-05. Default GROUP_1 (vence día 10, recargo día 11).
     */
    @Default
    @Enumerated(EnumType.STRING)
    @Column(name = "payment_group", nullable = false, length = 20)
    private PaymentGroup paymentGroup = PaymentGroup.GROUP_1;

    /**
     * Cómo se distribuyen los ítems del enrollment en el cronograma de cuotas.
     * Se persiste (V033) para que el auto-link a Presupuesto pueda separar el
     * pago en asientos por concepto (docx Jaque 2026-07-20 §Presupuesto).
     * Default SEPARATE (matrícula aparte, curso a cuotas, libros a book_sales).
     */
    @Default
    @Enumerated(EnumType.STRING)
    @Column(name = "distribution_mode", nullable = false, length = 20)
    private InstallmentDistribution distributionMode = InstallmentDistribution.SEPARATE;

    @Column(name = "moodle_status", length = 20)
    private String moodleStatus;

    @Column(name = "notes", columnDefinition = "text")
    private String notes;
}

File: frontend/src/types/enrollment.ts
import type { Instant, UUID } from './common'

// Refleja com.imedba.modules.enrollment.entity.EnrollmentStatus
export type EnrollmentStatus = 'ACTIVE' | 'SUSPENDED' | 'COMPLETED' | 'CANCELLED'

export const ENROLLMENT_STATUSES: EnrollmentStatus[] = [
  'ACTIVE',
  'SUSPENDED',
  'COMPLETED',
  'CANCELLED',
]

export const ENROLLMENT_STATUS_LABELS: Record<EnrollmentStatus, string> = {
  ACTIVE:    'Activa',
  SUSPENDED: 'Suspendida',
  COMPLETED: 'Completada',
  CANCELLED: 'Cancelada',
}

// Refleja com.imedba.common.enums.PaymentMethod
export type PaymentMethod =
  | 'TRANSFERENCIA'
  | 'EFECTIVO'
  | 'TARJETA_CREDITO'
  | 'TARJETA_DEBITO'
  | 'MERCADO_PAGO'
  | 'DEBITO_AUTOMATICO'
  | 'OTRO'

export const PAYMENT_METHODS: PaymentMethod[] = [
  'TRANSFERENCIA',
  'EFECTIVO',
  'TARJETA_CREDITO',
  'TARJETA_DEBITO',
  'MERCADO_PAGO',
  'DEBITO_AUTOMATICO',
  'OTRO',
]

export const PAYMENT_METHOD_LABELS: Record<PaymentMethod, string> = {
  TRANSFERENCIA:     'Transferencia',
  EFECTIVO:          'Efectivo',
  TARJETA_CREDITO:   'Tarjeta de crédito',
  TARJETA_DEBITO:    'Tarjeta de débito',
  MERCADO_PAGO:      'Mercado Pago',
  DEBITO_AUTOMATICO: 'Débito automático',
  OTRO:              'Otro',
}

// Refleja com.imedba.modules.enrollment.entity.PaymentGroup (reunión 2026-06-05).
// Grupo 1: cuota vence día 10, recargo día 11+. Grupo 2: vence día 20, recargo día 21+.
export type PaymentGroup = 'GROUP_1' | 'GROUP_2'

export const PAYMENT_GROUPS: PaymentGroup[] = ['GROUP_1', 'GROUP_2']

export const PAYMENT_GROUP_LABELS: Record<PaymentGroup, string> = {
  GROUP_1: 'Grupo 1 — vence día 10',
  GROUP_2: 'Grupo 2 — vence día 20',
}

export const PAYMENT_GROUP_SHORT: Record<PaymentGroup, string> = {
  GROUP_1: 'Grupo 1',
  GROUP_2: 'Grupo 2',
}

export function paymentGroupDueDay(g: PaymentGroup): number {
  return g === 'GROUP_2' ? 20 : 10
}

// Refleja EnrollmentResponse.StudentSummary
export interface EnrollmentStudentSummary {
  id:        UUID
  firstName: string
  lastName:  string
  email:     string
}

// Refleja EnrollmentResponse.CourseSummary
export interface EnrollmentCourseSummary {
  id:   UUID
  name: string
  code: string | null
}

// Refleja com.imedba.modules.enrollment.dto.EnrollmentResponse
export interface Enrollment {
  id:                 UUID
  student:            EnrollmentStudentSummary
  course:             EnrollmentCourseSummary
  discountCampaignId: UUID | null
  enrolledBy:         UUID | null
  enrollmentDate:     Instant

  listPrice:          number | null
  discountPercentage: number | null
  finalPrice:         number | null
  bookPrice:          number | null
  totalPrice:         number | null

  enrollmentFee:      number | null
  numInstallments:    number | null
  paymentGroup:       PaymentGroup

  contractFilePath:   string | null
  contractSentAt:     Instant | null
  contractSignedAt:   Instant | null

  status:             EnrollmentStatus
  moodleStatus:       string | null
  notes:              string | null

  createdAt:          Instant
  updatedAt:          Instant
}

// Refleja com.imedba.modules.enrollment.dto.EnrollmentCreateRequest
export interface EnrollmentCreateRequest {
  studentId:           UUID
  courseId:            UUID
  discountCampaignId?: UUID | null
  enrollmentDate?:     Instant | null

  listPrice?:          number | null
  discountPercentage?: number | null   // 0–100
  bookPrice?:          number | null

  enrollmentFee?:      number | null
  numInstallments?:    number | null   // ≥ 1
  paymentGroup?:       PaymentGroup    // default GROUP_1 en el backend
  useTotalDistribution?: boolean       // true = agrupar (curso+matrícula+libros / N); false = matrícula aparte

  contractFilePath?:   string | null   // max 500
  notes?:              string | null
}

// Refleja com.imedba.modules.enrollment.dto.EnrollmentUpdateRequest
// (no incluye studentId / courseId — inmutables después de crear)
export interface EnrollmentUpdateRequest {
  discountCampaignId?: UUID | null

  listPrice?:          number | null
  discountPercentage?: number | null
  bookPrice?:          number | null

  enrollmentFee?:      number | null
  numInstallments?:    number | null

  contractFilePath?:   string | null
  contractSentAt?:     Instant | null
  contractSignedAt?:   Instant | null

  notes?:              string | null
}
